domain/port/service/watcher.go

```diff
@@ -0,0 +1,166 @@
+	// for events based on the input change mask and mapper.
+	NewNamespaceMapperWatcher(
+		namespace string,
+		changeMask changestream.ChangeType,
+		initialStateQuery eventsource.NamespaceQuery,
+		mapper eventsource.Mapper,
+	) (watcher.StringsWatcher, error)
+
+	// NewNamespaceNotifyMapperWatcher returns a new namespace notify watcher
+	// for events based on the input change mask and mapper.
+	NewNamespaceNotifyMapperWatcher(
+		namespace string, changeMask changestream.ChangeType, mapper eventsource.Mapper,
+	) (watcher.NotifyWatcher, error)
+}
+
+// WatcherState describes the methods that the service needs for it's watchers.
+type WatcherState interface {
+	// WatchOpenedPortsTable returns the name of the table that should be watched
+	WatchOpenedPortsTable() string
+
+	// InitialWatchOpenedPortsStatement returns the query to load the initial event
+	// for the WatchOpenedPorts watcher
+	InitialWatchOpenedPortsStatement() string
+
+	// GetMachinesForEndpoints returns map from endpoint uuids to the uuids of
+	// the machines which host that endpoint for each provided endpoint uuid.
+	GetMachinesForEndpoints(ctx context.Context, endpointUUIDs []string) (map[string]string, error)
+
+	// GetApplicationsForEndpoints returns map from endpoint uuids to the uuids
+	// of the applications which host that endpoint for each provided endpoint uuid.
+	GetApplicationsForEndpoints(ctx context.Context, endpointUUIDs []string) (map[string]string, error)
+}
+
+// WatchOpenedPorts returns a strings watcher for opened ports. This watcher
+// will emit events for changes to the opened ports table. The events will
+// contain the machine UUIDs that are associated with the port range.
+func (s *WatchableService) WatchOpenedPorts(ctx context.Context) (watcher.StringsWatcher, error) {
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		eventsource.InitialNamespaceChanges(s.st.InitialWatchOpenedPortsStatement()),
+		s.portRangeToMachineMapper,
+	)
+}
+
+// WatchOpenedPortsForApplication returns a notify watcher for opened ports. This
+// watcher will emit events for changes to the opened ports table that are associated
+// with the given application
+func (s *WatchableService) WatchOpenedPortsForApplication(ctx context.Context, applicationUUID string) (watcher.NotifyWatcher, error) {
+	return s.watcherFactory.NewNamespaceNotifyMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		s.filterForApplication(applicationUUID),
+	)
+}
+
+// portRangeToMachineMapper is a mapper that maps port range events to machine UUIDs
+// of the machine hosting that endpoint
+func (s *WatchableService) portRangeToMachineMapper(
+	ctx context.Context, db database.TxnRunner, events []changestream.ChangeEvent,
+) ([]changestream.ChangeEvent, error) {
+
+	endpointUUIDs := transform.Slice(events, func(e changestream.ChangeEvent) string {
```

Comment:
I'm pretty sure the standard library offers this transform.Slice now.

Reply:
Are you sure? I'm under the impression the go devs specifically decided against it?

Reply:
It is a [proposal](https://github.com/golang/go/issues/61898), but not yet in the std.

However, iterators are a big enabler for this kind of function.

I proposed to use [samber/lo](https://github.com/samber/lo) for this kind of boilerplate few weeks ago, but we weren't into drag another dependency in juju codebase.